src/call/buffers.rs

```diff
@@ -0,0 +1,570 @@
+///
+/// Safety: see `VecSliceRefCount::new`, the same applies here.
+unsafe fn vec_slice(v: Vec<u8>) -> Box<grpc_slice> {
+    let mut data: MaybeUninit<grpc_sys::grpc_slice_grpc_slice_data> = MaybeUninit::zeroed();
+    *(*data.as_mut_ptr()).refcounted.as_mut() = grpc_sys::grpc_slice_grpc_slice_data_grpc_slice_refcounted {
+        bytes: v.as_ptr() as *const _ as *mut _,
+        length: v.len(),
+    };
+    let mut refcount = VecSliceRefCount::new(v);
+    let mut result = Box::new(grpc_slice {
+        refcount: ptr::null_mut(),
+        data: data.assume_init(),
+    });
+    refcount.slice = result.as_mut();
+    result.refcount = Box::into_raw(refcount) as *mut _;
+    result
+}
+
+/// A `grpc_slice_refcount` structure to handle ref-counting and memory mangement
+/// of `grpc_slice`s created by `vec_slice`.
+///
+/// Note that `grpc_slice_refcount` must be a prefix of this type so that a pointer
+/// to `VecSliceRefCount` can be treated polymorphically as a pointer to
+/// `grpc_slice_refcount`.
+#[repr(C)]
+struct VecSliceRefCount {
+    // 'vtable' pointer, should always point at VEC_SLICE_VTABLE.
+    vtable: *const grpc_slice_refcount_vtable,
+    // Self-reference. Must not be null.
+    sub_refcount: *mut grpc_slice_refcount,
+    // The `Vec` providing the memory for this slice.
+    vec: Vec<u8>,
+    // Refcount.
+    count: usize,
+    // Pointer to the slice this object is managing.
+    slice: *mut grpc_slice,
+    // Because `sub_refcount` is self-referential.
+    _phantom: PhantomPinned,
+}
+
+impl VecSliceRefCount {
+    /// Create a new `VecSliceRefCount`.
+    ///
+    /// Safety: the returned `VecSliceRefCount` has a 0 refcount. It is the
+    /// caller's responsibility to increment the refcount or either memory will
+    /// be leaked or `vec_slice_unref` will panic.
+    ///
+    /// `VecSliceRefCount::sub_refcount` is self-referential, therefore a
+    /// `VecSliceRefCount` *must never be moved*.
+    unsafe fn new(vec: Vec<u8>) -> Box<VecSliceRefCount> {
+        let mut result = Box::new(VecSliceRefCount {
+            vtable: &VEC_SLICE_VTABLE,
+            sub_refcount: ptr::null_mut(),
+            vec,
+            count: 0,
+            slice: ptr::null_mut(),
+            _phantom: PhantomPinned,
+        });
+        result.sub_refcount = &*result as *const _ as *mut _;
+        result
+    }
+}
+
+/// gRPC data structure proving functions for managing a `grpc_slice` created by
+/// `vec_slice`.
+static VEC_SLICE_VTABLE: grpc_slice_refcount_vtable = grpc_slice_refcount_vtable {
+    ref_: Some(vec_slice_ref),
+    unref: Some(vec_slice_unref),
+    eq: Some(grpc_sys::grpc_slice_default_eq_impl),
+    hash: Some(grpc_sys::grpc_slice_default_hash_impl),
+};
+
+/// Increment a vec_slice ref count.
+unsafe extern "C" fn vec_slice_ref(arg1: *mut ::std::os::raw::c_void) {
+    let refcount = arg1 as *mut VecSliceRefCount;
+    (*refcount).count += 1;
```

Comment:
I don't think it's safe. `grpc_slice` is expected to be shared between threads, and ref implement is expected to be Sync.

Reply:
Yes, I think you are right. `count` should be atomic and we should be using CAS, etc